test/chunked_stream_reader.dart

```diff
@@ -0,0 +1,360 @@
+      ['1', '2'],
+    ]));
+    expect(await s.readChunk(1), equals(['a']));
+    final sub = s.readStream(2);
+    await sub.drain();
+    expect(await s.readChunk(3), equals(['1', '2']));
+  });
+
+  test(
+      'readChunk() readStream() that ends with second chunk + '
+      '_readChunkedStream() readChunk()', () async {
+    final s = ChunkedStreamReader(_chunkedStream([
+      ['a', 'b', 'c'],
+      ['1', '2'],
+      ['3', '4']
+    ]));
+    expect(await s.readChunk(1), equals(['a']));
+    expect(
+        await s.readStream(4).toList(),
+        equals([
+          ['b', 'c'],
+          ['1', '2']
+        ]));
+    expect(await s.readChunk(3), equals(['3', '4']));
+  });
+
+  test(
+      'readChunk() readStream() that ends with second chunk + '
+      'drain() readChunk()', () async {
+    final s = ChunkedStreamReader(_chunkedStream([
+      ['a', 'b', 'c'],
+      ['1', '2'],
+      ['3', '4'],
+    ]));
+    expect(await s.readChunk(1), equals(['a']));
+    final substream = s.readStream(4);
+    await substream.drain();
+    expect(await s.readChunk(3), equals(['3', '4']));
+  });
+
+  test(
+      'readChunk() readStream() readChunk() before '
+      'draining substream produces StateError', () async {
+    final s = ChunkedStreamReader(_chunkedStream([
+      ['a', 'b', 'c'],
+      ['1', '2'],
+      ['3', '4'],
+    ]));
+    expect(await s.readChunk(1), equals(['a']));
+    // ignore: unused_local_variable
+    final substream = s.readStream(4);
+    expect(() async => await s.readChunk(3), throwsStateError);
+  });
+
+  test('creating two substreams simultaneously causes a StateError', () async {
+    final s = ChunkedStreamReader(_chunkedStream([
+      ['a', 'b', 'c'],
+      ['1', '2'],
+      ['3', '4'],
+    ]));
+    expect(await s.readChunk(1), equals(['a']));
+    // ignore: unused_local_variable
+    final substream = s.readStream(4);
+    expect(() async {
+      //ignore: unused_local_variable
+      final substream2 = s.readStream(3);
+    }, throwsStateError);
+  });
+
+  test('nested ChunkedStreamReader', () async {
+    final s = ChunkedStreamReader(_chunkedStream([
+      ['a', 'b'],
+      ['1', '2'],
+      ['3', '4'],
+    ]));
+    expect(await s.readChunk(1), equals(['a']));
+    final substream = s.readStream(4);
+    final nested = ChunkedStreamReader(substream);
+    expect(await nested.readChunk(2), equals(['b', '1']));
+    expect(await nested.readChunk(3), equals(['2', '3']));
+    expect(await nested.readChunk(2), equals([]));
+    expect(await s.readChunk(1), equals(['4']));
+  });
+
+  test('ChunkedStreamReaderByteStreamExt', () async {
+    final s = ChunkedStreamReader(_chunkedStream([
+      [1, 2, 3],
+      [4],
+    ]));
+    expect(await s.readBytes(1), equals([1]));
+    expect(await s.readBytes(1), isA<Uint8List>());
+    expect(await s.readBytes(1), equals([3]));
+    expect(await s.readBytes(1), equals([4]));
+    expect(await s.readBytes(1), equals([]));
+  });
+}
```

Comment:
Maybe do some tests with chunk size, and doing operations ending at chunk boundaries and not at chunk boundaries. 
And check cancelling a readStream right after the final chunk, at a chunk boundary, and see if I'm right it won't set `_reading` to false.

Reply:
You are absolute right, bug fixed -- thanks!